src/test/frontend/modules/outliner/core_test.cljs

```diff
@@ -310,26 +244,258 @@
+              :block/uuid 6}
+             {:db/id 9,
+              :block/left #:db{:id 6},
+              :block/level 3,
+              :block/parent #:db{:id 2},
+              :block/uuid 9}])
+           [{:db/id 6,
+             :block/left #:db{:id 3},
+             :block/level 1,
+             :block/parent #:db{:id 2},
+             :block/uuid 6}
+            {:db/id 9,
+             :block/left #:db{:id 6},
+             :block/level 1,
+             :block/parent #:db{:id 2},
+             :block/uuid 9}]))
+    (is (= (outliner-core/blocks-with-level
+            [{:db/id 6,
+              :block/left #:db{:id 3},
+              :block/level 3,
+              :block/parent #:db{:id 2},
+              :block/uuid 6}
+             {:db/id 9,
+              :block/left #:db{:id 6},
+              :block/level 4,
+              :block/parent #:db{:id 6},
+              :block/uuid 9}])
+           [{:db/id 6,
+             :block/left #:db{:id 3},
+             :block/level 1,
+             :block/parent #:db{:id 2},
+             :block/uuid 6}
+            {:db/id 9,
+             :block/left #:db{:id 6},
+             :block/level 2,
+             :block/parent #:db{:id 6},
+             :block/uuid 9}]))))
+
+;;; Fuzzy tests
+
+(def init-id (atom 100))
+
+(def unique-id (gen/fmap (fn [_] (swap! init-id inc)) gen/nat))
+(def compound (fn [inner-gen]
+                (gen/tuple unique-id (gen/vector inner-gen 1 2))))
+
+(def gen-node (gen/recursive-gen compound unique-id))
+
+(def gen-tree (gen/vector gen-node))
+
+(defn gen-blocks
+  []
+  (let [tree (->> (gen/generate gen-tree)
+                  (remove integer?))]
+    (when (seq tree)
+      (build-blocks tree))))
+
+(defn get-datoms
+  []
+  (d/datoms (db/get-db test-db) :avet :block/uuid))
+
+(defn get-random-block
+  []
+  (let [datoms (->> (get-datoms)
+                    (remove (fn [datom] (= 1 (:e datom)))))]
+    (if (seq datoms)
+      (let [id (:e (gen/generate (gen/elements datoms)))]
+        (db/pull test-db '[*] id))
+      (println "Empty DB so that there's no random blocks yet"))))
+
+(defn get-random-successive-blocks
+  []
+  (let [limit (inc (rand-int 20))]
+    (when-let [block (get-random-block)]
+      (loop [result [block]
+             node block]
+        (if-let [next (outliner-core/get-right-sibling (:db/id node))]
+          (let [next (db/pull test-db '[*] (:db/id next))]
+            (if (>= (count result) limit)
+              result
+              (recur (conj result next) next)))
+          result)))))
+
+(defn insert-blocks!
+  [blocks target]
+  (outliner-tx/transact! {:repo test-db}
+    (outliner-core/insert-blocks! blocks
+                                  target
+                                  {:sibling? (gen/generate gen/boolean)
+                                   :keep-uuid? true
+                                   :replace-empty-target? false})))
+
+(deftest random-inserts
+  (testing "Random inserts"
+    (transact-tree! tree)
```

Comment:
can we create a random init-tree to run fuzz-test?

Reply:
https://github.com/logseq/logseq/pull/4880/commits/4a5b227613ba9dbb92ee850a5ea6a7c9e3b4fac7#diff-4a0f4a15e5ca8a3981de57507a2c55189c307ce911173ccb6e384b8d78d1414bR382 switched to use a random tree.